docs/man/man3/BmsDeviceReader.3
.TH "BmsDeviceReader" 3 "Wed Jan 11 2017" "Version 0.0" "RaceUp_ino_core" \" -*- nroff -*-
.ad l
.nh
.SH NAME
BmsDeviceReader \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <BmsDeviceReader\&.h>\fP
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static byte * \fBbmsDeviceRead\fP (byte deviceAddress, byte regAddress, byte length)"
.br
.ti -1c
.RI "static double \fBgetTemperature\fP (byte device_address, byte temperature)"
.br
.ti -1c
.RI "static double \fBgetGpaiVbatt\fP (byte device_address, bool dev)"
.br
.ti -1c
.RI "static int \fBgetCellVoltage\fP (byte deviceAddress, byte cellNumber)"
.br
.ti -1c
.RI "static byte * \fBgetBmsDeviceStatus\fP (byte device_address)"
.br
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static const double \fBBETA_THERMISTOR\fP = 3435\&.0"
.br
.ti -1c
.RI "static const double \fBRB_THERMISTOR\fP = 1800\&.0"
.br
.ti -1c
.RI "static const double \fBRT_THERMISTOR\fP = 1500\&.0"
.br
.ti -1c
.RI "static const double \fBR0_THERMISTOR\fP = 10000\&.0"
.br
.in -1c
.SH "Detailed Description"
.PP 
Copyright 2016 - 2017 RaceUp Team Unipd
.PP
Licensed under the Apache License, Version 2\&.0 (the 'License'); you may not use this file except in compliance with the License\&. You may obtain a copy of the License at
.PP
http://www.apache.org/licenses/LICENSE-2.0
.PP
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied\&. See the License for the specific language governing permissions and limitations under the License\&. 
.SH "Member Function Documentation"
.PP 
.SS "byte * BmsDeviceReader::bmsDeviceRead (byte deviceAddress, byte regAddress, byte length)\fC [static]\fP"
Read data read from registers 
.PP
\fBParameters:\fP
.RS 4
\fIdeviceAddress\fP where to read 
.br
\fIregAddress\fP registry address 
.br
\fIlength\fP length to read 
.RE
.PP
\fBReturns:\fP
.RS 4
data read from registers 
.RE
.PP

.SS "byte * BmsDeviceReader::getBmsDeviceStatus (byte device_address)\fC [static]\fP"
Read status 
.PP
\fBParameters:\fP
.RS 4
\fIdeviceAddress\fP address of device to get info of 
.RE
.PP

.SS "int BmsDeviceReader::getCellVoltage (byte deviceAddress, byte cellNumber)\fC [static]\fP"
Return cell voltage 
.PP
\fBParameters:\fP
.RS 4
\fIdeviceAddress\fP address of device to get info of 
.br
\fIcellNumber\fP cell number 
.RE
.PP
\fBReturns:\fP
.RS 4
cell voltage of cell in device 
.RE
.PP

.SS "double BmsDeviceReader::getGpaiVbatt (byte device_address, bool dev)\fC [static]\fP"
TODO: WTF?? 
.PP
\fBParameters:\fP
.RS 4
\fIdevice_address\fP address of device 
.br
\fIdev\fP TODO: WTF?? 
.RE
.PP
\fBReturns:\fP
.RS 4
TODO: WTF?? 
.RE
.PP

.SS "double BmsDeviceReader::getTemperature (byte device_address, byte temperature)\fC [static]\fP"

.PP
\fBParameters:\fP
.RS 4
\fIdevice_address\fP address of device 
.br
\fItemperature\fP registry address of temperature to read 
.RE
.PP
\fBReturns:\fP
.RS 4
temperature of device 
.RE
.PP

.SH "Member Data Documentation"
.PP 
.SS "const double BmsDeviceReader::BETA_THERMISTOR = 3435\&.0\fC [static]\fP"
Thermistor data
.PP
Copyright 2016 - 2017 RaceUp Team Unipd
.PP
Licensed under the Apache License, Version 2\&.0 (the 'License'); you may not use this file except in compliance with the License\&. You may obtain a copy of the License at
.PP
http://www.apache.org/licenses/LICENSE-2.0
.PP
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied\&. See the License for the specific language governing permissions and limitations under the License\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for RaceUp_ino_core from the source code\&.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## 0.2.1 - 2017-01-15
### Fixed
- non-OOP residue includes

## 0.2 - 2017-01-15
### Refactored
- OOP design implemented
### Removed
- useless classes like BmsDeviceReader, BmsDeviceWriter BmsDeviceUpdater

## 0.1.3 - 2017-01-11
### Added
- utils frame conversions

## 0.1.2 - 2017-01-11
### Added
- SdCardStream
### Tested
- compilation with Arduino UNO, DUE

## 0.1.1 - 2017-01-11
### Fixed
- ```#include``` bugs

## 0.1 - 2017-01-11
### Added
- Utils, BmsDevice, BmsDeviceReader, BmsDeviceUpdater, BmsDeviceWriter classes
- documentation
- data

## 0.0 - 2017-01-11
### Added
- project structure and repository


src/Bms/BmsDeviceReader.cpp
/**
 * Copyright 2016 - 2017 RaceUp Team Unipd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


#include "BmsDeviceReader.h"
#include "../Utils.h"

#include "../data.h"

const double BmsDeviceReader::BETA_THERMISTOR = 3435.0;
const double BmsDeviceReader::RB_THERMISTOR = 1800.0;
const double BmsDeviceReader::RT_THERMISTOR = 1500.0;
const double BmsDeviceReader::R0_THERMISTOR = 10000.0;

byte *BmsDeviceReader::bmsDeviceRead(byte
deviceAddress,
byte regAddress, byte
length) {
#ifdef ARDUINO_UNO
SPI.setDataMode(SPI_MODE1);
byte logicalAddress = deviceAddress << 1;  // Shift the device bit and clear bit 0
logicalAddress &= 0b11111110;

byte receivedData[20];  // Create buffer for receivedData and clear it
memset(receivedData, 0, sizeof(receivedData));

digitalWrite(SLAVE_SELECT_PIN, LOW);  // take the SS pin low to select the chip:
delayMicroseconds(5);

SPI.transfer(logicalAddress);  // Send and receive SPI
SPI.transfer(regAddress);
SPI.transfer(length);
delayMicroseconds(5);

for (int i = 0; i <= length; i++) {
    receivedData[i] = SPI.transfer(0x00);
}
delayMicroseconds(5);
digitalWrite(SLAVE_SELECT_PIN,HIGH);

SPI.setDataMode(SPI_MODE0);
return receivedData;

#endif

#ifdef ARDUINO_DUE
// Take the SS pin low to select the chip
digitalWrite(SLAVE_SELECT_PIN,LOW);

delayMicroseconds(5);

//SPI.setDataMode(SPI_MODE1);

SPI.beginTransaction(SPISettings(50000, MSBFIRST, SPI_MODE1));

// Shift the device bit and clear bit 0
byte logicalAddress = deviceAddress << 1;
//logicalAddress &= ~(1 << 0);  // Clear bit 0, test bitClear(x,n)
logicalAddress &= 0b11111110;

// Create buffer for receivedData and clear it
byte receivedData[20];
memset(receivedData, 0, sizeof(receivedData));

//  Send and receive SPI
SPI.transfer(logicalAddress);
SPI.transfer(regAddress);
SPI.transfer(length);
delayMicroseconds(1);
for (int i = 0; i < length + 1; i++) {
    receivedData[i] = SPI.transfer(0x00);
}

delayMicroseconds(5);

// Take the SS pin high to de-select the chip
digitalWrite(SLAVE_SELECT_PIN,HIGH);


//SPI.setDataMode(SPI_MODE0);
SPI.endTransaction();

// Return data read from registers
return receivedData;
#endif

return 0;
}

double BmsDeviceReader::getTemperature(byte
device_address,
byte temperature
) {
byte *value_read = bmsDeviceRead(device_address, temperature, 2);  // read only first 2 bytes
byte tempRaw = (value_read[0] * 0x100) + value_read[1];  // multiply by 256 and add second byte
double value = (tempRaw + 2) / 33046.0;
value = ((1 / value) - 1) * BmsDeviceReader::RB_THERMISTOR - BmsDeviceReader::RT_THERMISTOR;
value = BmsDeviceReader::BETA_THERMISTOR /
        (log(value / (BmsDeviceReader::R0_THERMISTOR * exp(-BmsDeviceReader::BETA_THERMISTOR / 298.15))));
return
Utils::fromKelvinToCelsius(value);
}

double BmsDeviceReader::getGpaiVbatt(byte
device_address,
bool dev
) {
byte *gpaiRaw = bmsDeviceRead(device_address, GPAI, 0x02);

if (dev) {
return ((gpaiRaw[0] * 256) + gpaiRaw[1]) * (100.0 / 3.0) / 16383; // [V]
} else {
return ((gpaiRaw[0] * 256) + gpaiRaw[1]) * 2500.0 / 16383; // [mV]
}
}

int BmsDeviceReader::getCellVoltage(byte
deviceAddress,
byte cellNumber
) {
byte *cellRaw = bmsDeviceRead(deviceAddress, cellNumber, 0x02);
delay(ONE_CENTSECOND);

double parsed = Utils::convertCellVoltage((cellRaw[0] * 256) + cellRaw[1]);
return (int)
parsed;
}

byte *BmsDeviceReader::getBmsDeviceStatus(byte
device_address) {
return
bmsDeviceRead(device_address, DEVICE_STATUS,
1);  // read only 1 bit
}
